File: solutions/elixir/etl/1/lib/etl.ex
defmodule ETL do
  @doc """
  Transform an index into an inverted index.

  ## Examples

  iex> ETL.transform(%{"a" => ["ABILITY", "AARDVARK"], "b" => ["BALLAST", "BEAUTY"]})
  %{"ability" => "a", "aardvark" => "a", "ballast" => "b", "beauty" =>"b"}
  """
  @spec transform(map) :: map
  def transform(input) do
    input
    |> Enum.reduce([], fn {key, values}, acc ->
      [ (for v <- values, do: %{String.downcase(v) => key}) | acc ]
    end)
    |> List.flatten()
    |> Enum.reduce(%{}, fn pair, acc -> Map.merge(acc, pair) end)
  end
end

File: solutions/erlang/etl/1/src/etl.erl
-module(etl).

-export([transform/1, test_version/0]).

transform(Data) ->
	lists:flatten(
		lists:map(
			fun ({S, Chars}) -> transform(Chars, S, []) end,
			Data
		)
	).

transform([], _, Acc) -> Acc;
transform([[C]|More], S, Acc) when C>=$a andalso C=<$z -> transform(More, S, [{[C], S}|Acc]);
transform([[C]|More], S, Acc)-> transform(More, S, [{[C-$A+$a], S}|Acc]).

test_version() -> 1.